### PR TITLE
Commands GET `cat/aliases` and POST `_rollover` will work on specified indices

### LINKED ISSUE #1861: [BUG] Certain index APIs fail with permission denied when user role is restricted based on index pattern
**What is the bug?**
Certain index APIs fail with permission denied when user role is restricted based on index pattern

**How can one reproduce the bug?**
1. Create an internal role like so
`
curl -XPUT https://localhost:9200/_plugins/_security/api/roles/sem-role -H 'Content-Type: application/json' -d' { "cluster_permissions": [ "cluster_monitor" ], "index_permissions": [{ "index_patterns": [ "sem*" ], "allowed_actions" : ["*"] }] }
`
2. Create a user with this role
`
curl -XPUT https://localhost:9200/_plugins/_security/api/internalusers/sem-user -H 'Content-Type: application/json' -d' { "password": "######", "opendistro_security_roles": ["sem-role"] } '
`
3. Following APIs fail 
- Get aliases/alias, e.g. `curl https://localhost:9200/_cat/aliases/sem*?pretty -u sem-user` fails with _no permissions for [indices:admin/aliases/get]_
- Index rollover, e.g. `# curl -u sem-user -X POST "https://localhost:9200/sem-source1/_rollover?pretty"` fails with _no permissions for [indices:admin/rollover]_ 

These APIs work fine when the index pattern on the role is changed to "*" instead of "sem*"

**What is the expected behavior?**
- Get aliases/alias - should fetch aliases for index patterns the user has access to instead of throwing error
- Rollover should work fine given the user has permission to access the source index and create the target index based on the allowed index patterns for the user.

**What is your host/environment?**
OpenSearch 1.2.4 with opensearch-security plugin enabled on OEL7

**Do you have any screenshots?**
NA

**Do you have any additional context?**
- https://forum.opensearch.org/t/unable-to-set-permissions-for-put-index-template/4863/12
- https://forum.opensearch.org/t/internal-user-not-able-to-perform-operations-based-on-internal-role/9529 and https://forum.opensearch.org/t/internal-user-not-able-to-perform-operations-based-on-internal-role/9529/8

### PR DESCRIPTION
**Pull request by derek-ho**
### Description
Updates Get Aliases Request and Rollover to correctly look at the indices specified for these requests. This means that they will now work when a user has access to some specified set of indices, whereas before they would fail unless a user had permissions on all indices.

### Issues Resolved
#1861 

Is this a backport? If so, please add backport PR # and/or commits #

### Testing
[Please provide details of testing done: unit testing, integration testing and manual testing]

### Check List
- [ ] New functionality includes testing
- [ ] New functionality has been documented
- [ ] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/main/java/org/opensearch/security/privileges/PrivilegesEvaluator.java [hunk 1]
```diff
@@ -108,7 +108,8 @@ public class PrivilegesEvaluator {
             "indices:admin/shards/search_shards",
             "indices:admin/resolve/index",
             "indices:monitor/settings/get",
-            "indices:monitor/stats"
+            "indices:monitor/stats",
+            "indices:admin/aliases/get"
```
Comment (derek-ho) [inline src/main/java/org/opensearch/security/privileges/PrivilegesEvaluator.java:112]: It seems to me that the get aliases should also be a DNFOF like behavior - should I keep this in the PR and add a test for it?

File: src/integrationTest/java/org/opensearch/security/DoNotFailOnForbiddenTests.java [hunk 2]
```diff
@@ -427,4 +428,22 @@ public void shouldPerformCatIndices_positive() throws IOException {
...
+            // High level client doesn't support _cat/_indices API
+            Response getAliasesResponse = restHighLevelClient.getLowLevelClient().performRequest(getAliasesRequest);
+            List<String> aliases = new BufferedReader(new InputStreamReader(getAliasesResponse.getEntity().getContent())).lines()
+                .collect(Collectors.toList());
+
+            // Does not fail on forbidden, but alias response only contains index which user has access to
+            assertThat(getAliasesResponse.getStatusLine().getStatusCode(), equalTo(200));
+            assertThat(aliases.size(), equalTo(1));
+            assertThat(aliases.get(0), containsString("marvelous_songs"));
+            assertThat(aliases.get(0), not(containsString("horrible_songs")));
```
Comment (derek-ho) [inline src/integrationTest/java/org/opensearch/security/DoNotFailOnForbiddenTests.java:444]: @peternied @cwperks can you folks confirm this is the expected behavior (DNFOF returns alias even when user doesn't have index permissions on that alias when that alias contains an index that that user has access to (but the alias doesn't reveal any information about indices the user doesn't have access to))?

File: src/main/java/org/opensearch/security/resolver/IndexResolverReplacer.java [hunk 3]
```diff
@@ -750,11 +752,16 @@ private boolean getOrReplaceAllIndices(final Object request, final IndicesProvid
             }
             ((IndexRequest) request).index(newIndices.length != 1 ? null : newIndices[0]);
         } else if (request instanceof Replaceable) {
-            String[] newIndices = provider.provide(((Replaceable) request).indices(), request, true);
-            if (checkIndices(request, newIndices, false, allowEmptyIndices) == false) {
+            String[] newIndicesOrAliases;
+            if (request instanceof GetAliasesRequest) {
```
Comment (cwperks) [inline src/main/java/org/opensearch/security/resolver/IndexResolverReplacer.java:756]: I see that [GetAliasesRequest implements AliasesRequest](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/action/admin/indices/alias/get/GetAliasesRequest.java#L49). Can this use the more generic AliasesRequest? What other AliasesRequests are there and would they be impacted by the same issue?

Comment (derek-ho) [reply src/main/java/org/opensearch/security/resolver/IndexResolverReplacer.java:756]: Looks like the only other AliasesRequest I could find is within the IndicesAliasRequest: https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/action/admin/indices/alias/IndicesAliasesRequest.java#L100. From the description it is: `A request to add/remove aliases for one or more indices.`. IMO we can change this to `instaceof AliasesRequest`, wdyt?

Comment (cwperks) [reply src/main/java/org/opensearch/security/resolver/IndexResolverReplacer.java:756]: `AliasesRequests` extends `IndicesRequest.Replaceable` which means an `AliasesRequest` is also an `IndicesRequest`. In that case would the `provider.provide` call need to consider both `.aliases()` and `.indices()` from the request object?

Is it possible that a GetAliasesRequest can contain `.aliases()` and `.indices()` at the same time?

File: src/integrationTest/java/org/opensearch/security/DoNotFailOnForbiddenTests.java [hunk 4]
```diff
@@ -427,4 +428,22 @@ public void shouldPerformCatIndices_positive() throws IOException {
         }
     }
 
+    @Test
+    public void shouldPerformCatAliases_positive() throws IOException {
+        try (RestHighLevelClient restHighLevelClient = cluster.getRestHighLevelClient(LIMITED_USER)) {
+            Request getAliasesRequest = new Request("GET", "/_cat/aliases");
+            // High level client doesn't support _cat/_indices API
```
Comment (peternied) [inline src/integrationTest/java/org/opensearch/security/DoNotFailOnForbiddenTests.java:435]: This comment isn't correct, let's remove it.

If you feel strongly that it is 'weird' this API isn't on the high-level client API, I'd recommend opening an issue.

File: src/integrationTest/java/org/opensearch/security/DoNotFailOnForbiddenTests.java (hunk 2 above)
Comment (peternied) [reply src/integrationTest/java/org/opensearch/security/DoNotFailOnForbiddenTests.java:444]: Index names themselves are considered protected by the access control systems, so an alias that exposes an index name would violate the access control checks.  Therefore; Users see aliases for indices I they have access to see, and the inverse if the user does not have access to an index they should not see any aliases associated with it.

File: src/main/java/org/opensearch/security/resolver/IndexResolverReplacer.java (hunk 3 above)
Comment (DarshitChanpura) [reply src/main/java/org/opensearch/security/resolver/IndexResolverReplacer.java:756]: +1 . Will there be a case where indices() and aliases() are present at the same time. Which should be given higher priority in that case or should they both be considered ?

File: src/main/java/org/opensearch/security/resolver/IndexResolverReplacer.java [hunk 5]
```diff
@@ -781,6 +788,8 @@ private boolean getOrReplaceAllIndices(final Object request, final IndicesProvid
         } else if (request instanceof ReindexRequest) {
             result = getOrReplaceAllIndices(((ReindexRequest) request).getDestination(), provider, false) && result;
             result = getOrReplaceAllIndices(((ReindexRequest) request).getSearchRequest(), provider, false) && result;
+        } else if (request instanceof RolloverRequest) {
```
Comment (DarshitChanpura) [inline src/main/java/org/opensearch/security/resolver/IndexResolverReplacer.java:791]: should admins be allowed to rollover all indices with `*` permission or should these be index pattern specific? I'm not aware of any risks in allowing admin to rollover all indices, but would like to confirm.

**Comment by peternied** [comment]
@derek-ho looks like this PR has stalled, would you like to keep working on it, or would you prefer to close it out?

**Comment by peternied** [comment]
Oops meant to hit comment, not close.  I've reopened - let us know what you are thinking - thanks!

**Comment by derek-ho** [comment]
@peternied @DarshitChanpura closing this until I get more bandwidth to see it through to completion. Will keep the branch in its current state for now

**Pull request closed without merging**